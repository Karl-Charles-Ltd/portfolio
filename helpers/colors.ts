/**
 * Get the hex color codes that are used throughout the project.
 * These colors and names should match with
 * assets/scss/settings/_settings.colors.scss.
 *
 * @param {String} colorName - The color name from the theme
 * @returns {String} - The hex code for the matching color
 */
export function colorFromTheme(colorName) {
  const colors = {
    primary: '#370b3c',
    secondary: '#6b2f93',
    tertiary: '#D0438F',
    quaternary: '#f40ee3',
    quinary: '#09ECFD',
    'quinary-light': '#1A5C7E',
    'quinary-dark': '#010226',
    black: '#000000',
    background: '#202227',
    charcoal: '#4D4D4D',
    'grey-one': '#949494',
    'grey-two': '#CDCDCD',
    'light-grey': '#E2E2E2',
    'light-grey-one': '#E2E2E2',
    'light-grey-two': '#F9F9F9',
    white: '#FFFFFF',
  };

  return colors[colorName];
}

export const themeMap = {
  primary: {
    background: 'primary',
    foreground: 'secondary',
  },
  'primary-dark': {
    background: 'primary-dark',
    foreground: 'secondary',
  },
  secondary: {
    background: 'secondary',
    foreground: 'white',
    accent: 'tertiary',
  },
  'secondary-dark': {
    background: 'secondary-dark',
    foreground: 'white',
  },
  'secondary-light': {
    background: 'secondary-light',
    foreground: 'white',
  },
  tertiary: {
    background: 'tertiary',
    foreground: 'secondary',
  },
  quaternary: {
    background: 'quaternary',
    foreground: 'secondary',
  },
  quinary: {
    background: 'quinary',
    foreground: 'white',
  },
  'quinary-darker': {
    background: 'quinary-darker',
    foreground: 'white',
  },
  white: {
    background: 'white',
    foreground: 'secondary',
  },
  charcoal: {
    background: 'charcoal',
    foreground: 'white',
  },
  'light-grey-one': {
    background: 'light-grey-one',
    foreground: 'secondary',
  },
};
