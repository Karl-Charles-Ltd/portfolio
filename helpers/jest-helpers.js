import { shallowMount, createLocalVue } from '@vue/test-utils';
import { createRenderer } from 'vue-server-renderer';
import VueRouter from 'vue-router';

export function setup() {
  const renderer = createRenderer();
  const localVue = createLocalVue();
  const router = new VueRouter();

  // setupPlugins(useVuex, localVue);

  return {
    renderer,
    router,
    localVue,
    shallowMount,
  };
}

export function componentFactory(component, options) {
  return shallowMount(component, options);
}

export function createSnapshot(component, options) {
  const wrapper = componentFactory(component, options);
  expect(wrapper.exists()).toBeTruthy();
  expect(wrapper.html()).toMatchSnapshot();
}
