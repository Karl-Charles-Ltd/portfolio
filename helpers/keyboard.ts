export function addKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}

export function removeKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.removeEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}
