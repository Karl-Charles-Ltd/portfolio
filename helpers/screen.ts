import * as screenBreakpoints from '~/constants/screen-breakpoints';

export const breakpoints: string[] = [
  screenBreakpoints.MOBILE,
  screenBreakpoints.TABLET,
  screenBreakpoints.SMALL_DESK,
  screenBreakpoints.LARGE_DESK,
  screenBreakpoints.HUGE_DESK,
];

interface Screen {
  [key: string]: string | null;
}

export default function mq({ from = null, until = null }: Screen, currentScreenSize: string): boolean {
  const currentBreakpointIndex = breakpoints.indexOf(currentScreenSize);

  if (from !== '' && from !== null) {
    const fromBreakpointIndex = breakpoints.indexOf(from || '');
    if (fromBreakpointIndex > currentBreakpointIndex) {
      return false;
    }
  }

  if (until !== '' && until !== null) {
    const untilBreakpointIndex = breakpoints.indexOf(until || '');
    if (untilBreakpointIndex <= currentBreakpointIndex) {
      return false;
    }
  }

  return true;
}
