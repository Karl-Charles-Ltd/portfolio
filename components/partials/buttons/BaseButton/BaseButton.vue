<template>
  <component
    :is="tag"
    :type="nativeType"
    :class="[buttonClasses, additionalButtonClasses]"
    v-bind="$attrs"
    v-on="buttonListeners"
  >
    <span v-if="active && type === 'outline'" class="v-btn__left-content">
      <BaseIcon icon="Tick" :theme="themeMap[theme].foreground" size="base" />
    </span>
    <span v-else-if="$slots['left-content']" class="v-btn__left-content">
      <slot name="left-content" />
    </span>

    <BaseIcon
      v-if="loading"
      icon="LoadingSpinner"
      :theme="loadingSpinnerColor"
      :size="type === 'outline' || type === 'action' ? 'base' : 'small'"
    />

    <span v-else :class="contentClasses" class="v-btn__content">
      <slot />
    </span>
  </component>
</template>

<script lang="ts">
import { Prop, Component, mixins } from 'nuxt-property-decorator';
import buttonMixin from '../buttonMixin.vue';
import { themeMap } from '~/helpers/colors';

const availableTypes: string[] = ['block', 'outline', 'line', 'action'];

/**
 * ### **Features:**
 * - Themes - primary, secondary, tertiary, quaternary, quinary, white, charcoal, light-grey-one
 * - Types  - block, outline, line, action
 * - States - disabled, loading, active
 * - Tag    - button, a, nuxt-link
 *
 * ### **Important attributes:**
 * `aria-label` - Add an aria-label when button content doesn't describe it, such as when icon-only. Read as `"click {label}"`
 * `aria-role` - When not using `<button>` tag, add `aria-role="button"` if the element will do something in response to user input (such as when using an a tag to display a modal, rather than to navigate).
 *
 * ### **Other notes:**
 * Block and outline styles - the width is set to 100% so that the parent can decide the width of the button.
 */
@Component
export default class BaseButton extends mixins(buttonMixin) {
  @Prop({ default: 'block', validator: (value) => availableTypes.includes(value) }) readonly type!: string;
  @Prop({ default: false }) readonly wide!: boolean;
  @Prop({ default: false }) readonly block!: boolean;

  themeMap = themeMap;

  get contentClasses() {
    return {
      'u-margin-left-micro': !!this.$slots['left-content'],
      'u-text--small-body': this.type === 'action',
      'u-text--button': this.type !== 'action',
    };
  }

  get additionalButtonClasses() {
    return {
      [`v-btn--${this.type}`]: this.type,
      'v-btn--has-left-content': !!this.$slots['left-content'],
      'v-btn--wide': this.wide,
      'v-btn--display-block': this.block,
    };
  }

  get loadingSpinnerColor(): string | null {
    if (this.type === 'outline' || this.type === 'line') return this.theme;
    if (this.type === 'action') return 'secondary';

    return themeMap[this.theme] ? themeMap[this.theme].foreground : null;
  }
}
</script>

<style lang="scss">
@import '../themes.scss';

.v-btn {
  -webkit-appearance: none !important;
  -moz-appearance: none !important;
  appearance: none !important;
  background: none;
  border: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  padding: 0;
  position: relative;
}

.v-btn--outline,
.v-btn--action {
  .v-btn__content,
  .v-btn__left-content {
    display: block;
    position: relative;
    z-index: z-index('foreground');
  }
}

.v-btn--block {
  padding: $spacing-tiny $spacing-teeny;
  position: relative;
  text-align: center;
  transition-duration: 0.3s;
  transition-property: transform;
  transition-timing-function: ease-in-out;

  @include mq($until: tablet) {
    min-width: 28rem;
  }

  @include mq($from: tablet) {
    padding: $spacing-tiny $spacing-intermediate;
  }

  @each $style-name, $style-map in $_block-button-style-map {
    $_other-styles: map-get($style-map, 'other');

    &.v-btn--#{$style-name} {
      @each $style, $style-value in $style-map {
        @if ($style != 'other') {
          #{$style}: #{$style-value};
        }
      }

      &:after {
        background-color: map-get($style-map, 'background-color');
        content: '';
        display: block;
        height: 0.8rem;
        left: 1%;
        position: absolute;
        top: 103%;
        transform: rotate3d(2, 0, 0, -0.2turn) rotateX(-45deg);
        transform-origin: top;
        transition-duration: 0.3s;
        transition-property: transform, background-color, height;
        transition-timing-function: ease-in-out;
        width: 98%;
      }

      &.v-btn--loading {
        $_loading-color: map-get($_other-styles, 'loading');
        &:after {
          border-color: transparent transparent #{$_loading-color} #{$_loading-color};
        }
      }

      &.v-btn--disabled {
        $_disabled-color: map-get($_other-styles, 'disabled');
        color: $_disabled-color;
      }

      &:hover:not(.v-btn--disabled):not(:active):not(.v-btn--no-hover) {
        transform: perspective(10rem) scale(1.01);
        &:after {
          background-color: map-get($_other-styles, 'box-shadow-color');
          transform: perspective(5rem) scale(1.05) rotateX(-45deg);
        }
      }

      &:focus:not(.v-btn--disabled) {
        border-color: transparent;
        outline-color: map-get($style-map, 'background-color');
      }
    }
  }
}

.v-btn--outline {
  align-items: center;
  display: flex;
  justify-content: center;
  padding: $spacing-mini $spacing-intermediate;
  position: relative;
  transition-duration: 0.3s;
  transition-property: color;
  transition-timing-function: ease-in-out;

  &.v-btn--active {
    .v-btn__left-content {
      left: $spacing-teeny;
      position: absolute;
    }
  }

  @include mq($until: tablet) {
    width: 100%;
  }

  @each $style-name, $style-map in $_outline-button-style-map {
    $_other-styles: map-get($style-map, 'other');
    $_background-color: map-get($_other-styles, 'selected-background-color');

    &.v-btn--#{$style-name} {
      @each $style, $style-value in $style-map {
        @if ($style != 'other') {
          #{$style}: #{$style-value};
        }
      }

      &:after {
        background-color: map-get($_other-styles, 'selected-background-color');
        content: '';
        display: block;
        height: 100%;
        left: 0;
        position: absolute;
        top: 0;
        transition-duration: 0.3s;
        transition-property: width;
        transition-timing-function: ease-in-out;
        width: 0;
        z-index: z-index('base');
      }

      &.v-btn--active {
        background-color: map-get($_other-styles, 'selected-background-color');
        color: map-get($_other-styles, 'hover-text-color');
      }

      &:hover:not(.v-btn--disabled):not(.v-btn--loading):not(.v-btn--active):not(.v-btn--no-hover) {
        color: map-get($_other-styles, 'hover-text-color');
        &:after {
          width: 100%;
        }
      }

      &:focus:not(.v-btn--disabled) {
        outline-color: map-get($_other-styles, 'selected-background-color');
      }
    }
  }
}

.v-btn--line {
  display: inline-block;
  padding: 0 0 $spacing-mini 0;
  position: relative;

  @each $style-name, $style-map in $_line-button-style-map {
    &.v-btn--#{$style-name} {
      $_button-color: map-get($style-map, 'color');
      color: $_button-color;

      &:before,
      &:after {
        background-color: $_button-color;
        bottom: 0;
        content: '';
        display: block;
        height: 3px;
        position: absolute;
        right: 0;
        transition: all ease(out-quad);
        width: 100%;
      }

      &:before {
        right: unset;
        transition-delay: 0s;
        transition-duration: 0.11s;
        width: 0;
      }
      &:after {
        transition-delay: 0.22s;
        transition-duration: 0.33s;
        width: 100%;
      }

      &:hover {
        transition-timing-function: ease(in-quad);
        &:before {
          transition-delay: 0.22s;
          transition-duration: 0.33s;
          width: 100%;
        }
        &:after {
          transition-delay: 0s;
          transition-duration: 0.11s;
          width: 0;
        }
      }

      &:focus:not(.v-btn--disabled) {
        outline-color: $_button-color;
      }
    }
  }
}

.v-btn--action {
  align-items: center;
  display: flex;
  justify-content: center;
  min-height: $spacing-mini;
  min-width: rem-value(106);
  padding: $spacing-mini $spacing-mini 0.5rem $spacing-mini;
  position: relative;
  transition: {
    duration: 0.3s;
    property: color;
    timing-function: ease-in-out;
  }

  .v-btn__content {
    line-height: 1;
  }

  &.v-btn--has-left-content,
  &.v-btn--loading {
    align-items: flex-end;
    padding: $spacing-mini;
  }

  @each $style-name, $style-map in $_action-button-style-map {
    $_other-styles: map-get($style-map, 'other');
    $_background-color: map-get($_other-styles, 'hover-background-color');

    &.v-btn--#{$style-name} {
      @each $style, $style-value in $style-map {
        @if ($style != 'other') {
          #{$style}: #{$style-value};
        }
      }

      &:after {
        background-color: $_background-color;
        content: '';
        display: block;
        height: 100%;
        left: 0;
        position: absolute;
        top: 0;
        transition: {
          duration: 0.3s;
          property: width;
          timing-function: ease-in-out;
        }
        width: 0;
        z-index: z-index('base');
      }

      .v-icon {
        transition: {
          duration: 0.2s;
          property: color;
          timing-function: ease-in-out;
        }
      }

      &:hover:not(.v-btn--disabled):not(.v-btn--loading):not(.v-btn--no-hover) {
        color: map-get($_other-styles, 'hover-text-color');
        &:after {
          width: 100%;
        }
        .v-btn__left-content .v-icon {
          color: map-get($_other-styles, 'hover-text-color') !important;
        }
      }

      &:focus:not(.v-btn--disabled) {
        outline-color: $_background-color;
      }
    }
  }
}

/* Modifiers
   ========================================================================== */

.v-btn--disabled {
  cursor: not-allowed;
  pointer-events: none;

  &.v-btn-icon,
  &.v-btn--line,
  &.v-btn--outline,
  &.v-btn--action,
  &.v-btn--block .v-btn__content {
    opacity: 0.5;
  }
}

.v-btn--loading {
  pointer-events: none;
  position: relative;
}

.v-btn--wide {
  &.v-btn--block,
  &.v-btn--outline {
    min-width: 28rem;
    @include mq($from: tablet) {
      min-width: 30rem;
    }
  }

  &.v-btn--block {
    padding: $spacing-small $spacing-teeny;
  }

  &.v-btn--outline {
    padding: $spacing-teeny $spacing-teeny;
  }
}

.v-btn--display-block {
  display: block;
  width: 100%;
}
</style>
