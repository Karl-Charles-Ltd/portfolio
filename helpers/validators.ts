import { themeMap } from './colors';

export function validateTheme(theme: string, extraColors: string[] = []): boolean {
  const baseThemes: string[] = Object.keys(themeMap);
  const availableThemes: string[] = baseThemes.concat(extraColors);

  return availableThemes.includes(theme);
}
