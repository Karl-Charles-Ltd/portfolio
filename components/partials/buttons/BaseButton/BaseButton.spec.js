import BaseButton from './BaseButton.vue';
import { setup, componentFactory, createSnapshot } from '~/helpers/jest-helpers';

const { localVue } = setup();

const shallowMountComponent = (options) => componentFactory(BaseButton, { localVue, ...options });

describe('BaseButton', () => {
  const stubs = ['BaseIcon'];
  test('mounts and renders properly', () => {
    createSnapshot(BaseButton, { localVue, stubs });
  });

  const buttonTypes = ['block', 'outline', 'line', 'action'];
  for (let test = 0; test < buttonTypes.length; test += 1) {
    const buttonType = buttonTypes[test];

    it(`has the ${buttonType} class if the type is ${buttonType}`, () => {
      const wrapper = shallowMountComponent({ propsData: { type: buttonType }, stubs });
      expect(wrapper.classes().includes(`v-btn--${buttonType}`)).toBeTruthy();
    });
  }

  const buttonStates = ['disabled', 'active', 'loading'];
  for (let test = 0; test < buttonStates.length; test += 1) {
    const buttonState = buttonStates[test];

    it(`has the ${buttonState} class if the state is ${buttonState}`, () => {
      const wrapper = shallowMountComponent({ propsData: { [buttonState]: true }, stubs });
      expect(wrapper.classes().includes(`v-btn--${buttonState}`)).toBeTruthy();
    });
  }
});
