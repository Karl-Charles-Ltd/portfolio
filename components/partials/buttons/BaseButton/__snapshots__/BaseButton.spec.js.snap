// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton mounts and renders properly 1`] = `
"<button type=\\"button\\" class=\\"v-btn v-btn--secondary v-btn--block\\">
  <!----> <span class=\\"v-btn__content u-text--button\\"></span></button>"
`;
